azkaban-common/src/main/java/azkaban/project/JdbcProjectHandlerSet.java

```diff
@@ -40,97 +41,100 @@
-        "SELECT id, name, active, modified_time, create_time, version, last_modified_by, description, enc_type, settings_blob FROM projects WHERE active=true";
+    // Still return the project if it has no associated permissions
+    public static final String SELECT_PROJECT_BY_ID = BASE_QUERY + "LEFT JOIN project_permissions prm ON prj.id = prm.project_id WHERE prj.id=?";
 
-    public static String SELECT_ACTIVE_PROJECT_BY_NAME =
-        "SELECT id, name, active, modified_time, create_time, version, last_modified_by, description, enc_type, settings_blob FROM projects WHERE name=? AND active=true";
+    // Still return the project if it has no associated permissions
+    public static final String SELECT_ACTIVE_PROJECT_BY_NAME = BASE_QUERY + "LEFT JOIN project_permissions prm ON prj.id = prm.project_id WHERE prj.name=? AND prj.active=true";
+
+    // ONLY return projects that have at least one associated permission, this is for performance reasons.
+    // (JOIN is way faster than LEFT JOIN)
+    public static final String SELECT_ALL_ACTIVE_PROJECTS = BASE_QUERY + "JOIN project_permissions prm ON prj.id = prm.project_id WHERE prj.active=true";
 
     @Override
     public List<Project> handle(final ResultSet rs) throws SQLException {
       if (!rs.next()) {
         return Collections.emptyList();
       }
 
-      final ArrayList<Project> projects = new ArrayList<>();
+      // Project ID -> Project
+      final Map<Integer, Project> projects = new HashMap<>();
       do {
         final int id = rs.getInt(1);
-        final String name = rs.getString(2);
-        final boolean active = rs.getBoolean(3);
-        final long modifiedTime = rs.getLong(4);
-        final long createTime = rs.getLong(5);
-        final int version = rs.getInt(6);
-        final String lastModifiedBy = rs.getString(7);
-        final String description = rs.getString(8);
-        final int encodingType = rs.getInt(9);
-        final byte[] data = rs.getBytes(10);
-
-        final Project project;
-        if (data != null) {
-          final EncodingType encType = EncodingType.fromInteger(encodingType);
-          final Object blobObj;
-          try {
-            // Convoluted way to inflate strings. Should find common package or
-            // helper function.
-            if (encType == EncodingType.GZIP) {
-              // Decompress the sucker.
-              final String jsonString = GZIPUtils.unGzipString(data, "UTF-8");
-              blobObj = JSONUtils.parseJSONFromString(jsonString);
-            } else {
-              final String jsonString = new String(data, "UTF-8");
-              blobObj = JSONUtils.parseJSONFromString(jsonString);
+
+        // If a project has multiple permissions - the project will be returned multiple times,
+        // one for each permission and we don't need to go through the work of reconstructing the
+        // project object if we've already seen it.
+        if (!projects.containsKey(id)) {
```

Comment:
No need to change anything now, but this could be simplified to just checking the last fetched ID against the previous one. I think we decided that the resultset would be sorted by ID due to it being [the prefix of] the PK on both tables. This would require an ORDER BY clause in the SELECT just because the order is not guaranteed otherwise, but that would be free given the implementation ...and everything we know :) .

Reply:
True...it _could_ :)